pkg/kubelet/volume/cache/actual_state_of_world.go

```diff
@@ -0,0 +1,586 @@
+	// If a pod with the same unique name does not exist under the specified
+	// volume, false is returned.
+	// If a volume with the name volumeName does not exist in the list of
+	// attached volumes, a volumeNotAttachedError is returned indicating the
+	// given volume is not yet attached.
+	// If a the given volumeName/podName combo exists but the value of
+	// remountRequired is true, a remountRequiredError is returned indicating
+	// the given volume has been successfully mounted to this pod but should be
+	// remounted to reflect changes in the referencing pod. Volumes like
+	// Downward API, depend on this to update the contents of the volume.
+	// All volume mounting calls should be idempotent so a second mount call for
+	// volumes that do not need to update contents should not fail.
+	PodExistsInVolume(podName volumetypes.UniquePodName, volumeName api.UniqueVolumeName) (bool, error)
+
+	// GetMountedVolumes generates and returns a list of volumes and the pods
+	// they are successfully attached and mounted for based on the current
+	// actual state of the world.
+	GetMountedVolumes() []MountedVolume
+
+	// GetMountedVolumesForPod generates and returns a list of volumes that are
+	// successfully attached and mounted for the specified pod based on the
+	// current actual state of the world.
+	GetMountedVolumesForPod(podName volumetypes.UniquePodName) []MountedVolume
+
+	// GetAttachedVolumes generates and returns a list of all attached volumes.
+	GetAttachedVolumes() []AttachedVolume
+
+	// GetUnmountedVolumes generates and returns a list of attached volumes that
+	// have no mountedPods. This list can be used to determine which volumes are
+	// no longer referenced and may be detached.
+	GetUnmountedVolumes() []AttachedVolume
+}
+
+// MountedVolume represents a volume that has successfully been mounted to a pod.
+type MountedVolume struct {
+	operationexecutor.MountedVolume
+}
+
+// AttachedVolume represents a volume that is attached to a node.
+type AttachedVolume struct {
+	operationexecutor.AttachedVolume
+
+	// GloballyMounted indicates that the volume is mounted to the underlying
+	// device at a global mount point. This global mount point must unmounted
+	// prior to detach.
+	GloballyMounted bool
+}
+
+// NewActualStateOfWorld returns a new instance of ActualStateOfWorld.
+func NewActualStateOfWorld(
+	hostName string,
+	volumePluginMgr *volume.VolumePluginMgr) ActualStateOfWorld {
+	return &actualStateOfWorld{
+		hostName:        hostName,
+		attachedVolumes: make(map[api.UniqueVolumeName]attachedVolume),
+		volumePluginMgr: volumePluginMgr,
+	}
+}
+
+// IsVolumeNotAttachedError returns true if the specified error indicates that
+// PodExistsInVolume() failed to find volume in the list of attached volumes.
+func IsVolumeNotAttachedError(err error) bool {
+	_, ok := err.(volumeNotAttachedError)
+	return ok
+}
+
+// IsRemountRequiredError returns true if the specified error indicates that
```

Comment:
I would change the doc for this to just be 'returns true if the specified error is a remountRequiredError'